zkevm-circuits/src/blake2b/circuit/src/lib.rs

```diff
@@ -0,0 +1,1855 @@
+    p(k) and then p(j). Since the value m permuted for the i-th round is (m[SIGMA[i % 10][0]], m[SIGMA[i % 10][1]], ..., m[SIGMA[i % 10][15]]), 
+    the value of SIGMA[i % 10][j] is replaced with j by p(i). On the other hand, per(SIGMA[i % 10]) replaces j with SIGMA[i % 10][j], therefore, 
+    p(i) is per'(SIGMA[i % 10]). So P(i) = per'(SIGMA[(i + 1) % 10]) * per(SIGMA[i % 10]). The permutations computed by this formula are used by 
+    RoundGate class gate to compute the values of m permuted for the next rounds without the need to access the value of m. 
+*/
+
+use std::{ marker::PhantomData, array, convert::TryInto };
+use halo2_proofs::{ arithmetic::FieldExt, circuit::*, plonk::*, poly::Rotation };
+
+// Blake2 initialization vector
+const IV: [u64; 8] = [0x6a09e667f3bcc908, 0xbb67ae8584caa73b, 0x3c6ef372fe94f82b, 0xa54ff53a5f1d36f1, 
+                      0x510e527fade682d1, 0x9b05688c2b3e6c1f, 0x1f83d9abfb41bd6b, 0x5be0cd19137e2179];
+
+// Blake2 message schedule
+const SIGMA:[[usize; 16]; 10] = [
+    [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+    [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+    [11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4],
+    [7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8],
+    [9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13],
+    [2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9],
+    [12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11],
+    [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
+    [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
+    [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0]];
+
+// Optimal numbers of rows per round for the compression circuit. Optimality means that every 
+// number is the smallest one among all that result in the same number of columns in the circuit 
+const ROWS_PER_ROUND:[usize; 12] = [8, 16, 24, 32, 40, 48, 56, 64, 80, 88, 120, 128];
+
+// A small module providing the implementation of the BLAKE2b compression function
+pub mod blake2b {
+    //  Computes the mixing function G. The elements of the input are specified according to RFC 7693. 
+    //  The computation results are returned by means of updating the vector referenced by "v" 
+    fn g(v: &mut [u64; 16], a: usize, b: usize, c: usize, d: usize, x: u64, y: u64) {
+        v[a] = v[a].wrapping_add(v[b]).wrapping_add(x);
+        v[d] = (v[d] ^ v[a]).rotate_right(32);
+    
+        v[c] = v[c].wrapping_add(v[d]);
+        v[b] = (v[b] ^ v[c]).rotate_right(24);
+    
+        v[a] = v[a].wrapping_add(v[b]).wrapping_add(y);
+        v[d] = (v[d] ^ v[a]).rotate_right(16); 
+        
+        v[c] = v[c].wrapping_add(v[d]);
+        v[b] = (v[b] ^ v[c]).rotate_right(63);
+    }
+    
+    // Computes the BLAKE2b compression function. The elements of the input are specified according to RFC 7693  
+    pub fn compress(r: u32, h: &[u64; 8], m: &[u64; 16], t: u128, f: bool) -> [u64; 8] {
```

Comment:
The circuit is not for BLAKE2, but for the BLAKE2 compression function `F` precompile, for which r is a part of the input.

Reply:
Do you think its possible to link the specs saying `r` is arbitrary here?

Reply:
Since the circuit is dedicated to this EIP-152 (https://eips.ethereum.org/EIPS/eip-152), but not to BLAKE2,I am sure it is not necessary. And even in RFC-7693 the number of rounds is designated as r and varies depending on the BLAKE2 versions.